include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h

```diff
@@ -57,6 +59,150 @@ struct __leaf_sort_kernel
+            const std::uint32_t __end1 = std::min<std::uint32_t>(__start1 + __sorted, __end);
+            const std::uint32_t __start2 = __end1;
+            const std::uint32_t __end2 = std::min<std::uint32_t>(__start2 + __sorted, __end);
+            const std::uint32_t __n1 = __end1 - __start1;
+            const std::uint32_t __n2 = __end2 - __start2;
+
+            const auto& __in = __storage.begin() + __data_in_temp * __sorted_final;
+            const auto& __out = __storage.begin() + (!__data_in_temp) * __sorted_final;
+            const auto& __in1 = __in + __start1;
+            const auto& __in2 = __in + __start2;
+
+            const auto __start = __find_start_point(__in1, __in2, __id_local, __n1, __n2, __comp);
+            // TODO: copy the data into registers before the merge to halve the required amount of SLM
+            __serial_merge(__in1, __in2, __out, __start.first, __start.second, __id, __data_per_workitem, __n1, __n2,
+                           __comp);
+            __dpl_sycl::__group_barrier(__item);
+
+            __sorted = __next_sorted;
+            __next_sorted *= 2;
+            __data_in_temp = !__data_in_temp;
+        }
+        return __data_in_temp;
+    }
+};
+
+template <std::uint16_t _DataPerWorkitem, std::uint16_t _WorkGroupSize, typename _Range, typename _Compare>
+struct __leaf_sorter
+{
+    static constexpr std::uint16_t __data_per_workitem = _DataPerWorkitem;
+    static constexpr std::uint16_t __workgroup_size = _WorkGroupSize;
+    static constexpr std::uint32_t __process_size = __data_per_workitem * __workgroup_size;
+
+    using _Tp = oneapi::dpl::__internal::__value_t<_Range>;
+    using _Size = oneapi::dpl::__internal::__difference_t<_Range>;
+    using _Storage = __dpl_sycl::__local_accessor<_Tp>;
+    // TODO: select a better sub-group sorter depending on sort stability,
+    //       a type (e.g. it can be trivially copied for shuffling within a sub-group)
+    using _SubGroupSorter = __subgroup_bubble_sorter;
+    using _GroupSorter = __group_merge_path_sorter<__data_per_workitem, __workgroup_size>;
+
+    constexpr static std::uint32_t
+    storage_size()
+    {
+        return 2 * __data_per_workitem * __workgroup_size;
+    }
+
+    void
+    initialize_storage(sycl::handler& __cgh)
+    {
+        __storage = _Storage(storage_size(), __cgh);
+    }
+
+    __leaf_sorter(_Range& __rng, _Compare __comp)
+        : __rng(__rng), __comp(__comp), __n(__rng.size()), __sub_group_sorter(), __group_sorter()
+    {
+    }
+
+    void
+    sort(const sycl::nd_item<1>& __item) const
+    {
+        sycl::sub_group __sg = __item.get_sub_group();
+        sycl::group __wg = __item.get_group();
+        const std::uint32_t __wg_id = __wg.get_group_linear_id();
+        const std::uint32_t __sg_id = __sg.get_group_linear_id();
+        const std::uint32_t __sg_size = __sg.get_local_linear_range();
+        const std::uint32_t __sg_inner_id = __sg.get_local_linear_id();
```

Comment:
```suggestion
        const std::uint32_t __sg_local_id = __sg.get_local_linear_id();
```
Should we change this to follow SYCL's language?

Reply:
Done